Warn about missing submodules

programonauta opened the pull request:
Related to #202.

For now just updated CMakeLists.txt on root directory

programonauta committed:
Test if sub-modules exist and give an error message if are missing

chriseth commented on CMakeLists.txt:
```diff
@@ -10,12 +19,22 @@ set(WEB3_DIR "${CMAKE_CURRENT_LIST_DIR}/webthree" CACHE PATH "The path to the we
 project(cpp-ethereum VERSION "1.2.1")
 
+# test if files exist
+testFileExists("${CMAKE_CURRENT_LIST_DIR}/webthree-helpers/cmake/EthCompilerSettings.cmake")
```
I think it's fine to test just a single file, perhaps even just check whether the 'cmake' directory is present? After all, this should only check whether we actually checked out the sub-repository or not.

programonauta replied:
Ok, but when git clone without --recursive the cmake directory is downloaded but empty. I noticed that test only the directory is not enough.

chriseth commented on CMakeLists.txt:
```diff
@@ -1,3 +1,12 @@
+function(testFileExists fileName)
+
+   if (NOT EXISTS ${fileName})
+      message(FATAL_ERROR "Could not find file " ${fileName})
```
Perhaps not at the global level, but certainly at the point where this is used, we should print a more detailed message like "Did you use 'git clone --recursive' to checkout all sub-repositories".

programonauta replied:
I decided create a function to test each file. Because after the second include, the cmake create a big number of messages without stop the process, and become very difficult to find out where exactly are the error message.

So, this function create a standard message, that could have more details creating a new parameter, and exit as a fatal error.

bobsummerwill commented on CMakeLists.txt:
```diff
@@ -10,12 +42,22 @@ set(WEB3_DIR "${CMAKE_CURRENT_LIST_DIR}/webthree" CACHE PATH "The path to the we
 # project name and version should be set after cmake_policy CMP0048
 project(cpp-ethereum VERSION "1.2.1")
 
+# test if files exist
```
I wouldn't bother with checking all of these individual files.     This whole "feature" is purely heuristic anyway, so if EthPolicy.cmake is there, it's safe enough to assume everything else is.

If we start checking every file then we "should" do that for every single include everywhere, which would obviously be stupid.

The heuristic check is just really to try to help the use if they forgot the --recursive.    Just the one check on the very first file should be enough for this, I think.   Thanks!    Let's chop these other ones out.

programonauta commented:
@bobsummerwill, I tested a `git clone` without `--recursive` option in a new directory and the cmake directory didn't come, only `webthree-helpers` come but empty.
So I just tested if `cmake` directory exists to find out if the user cloned all tree.

bobsummerwill commented:
Great - that makes sense.   Please could you strip the escape codes down to just the three you use?   Thanks!

programonauta commented:
@bobsummerwill done only with escape codes used on script.

I didn't test in Windows, but I tested the behavior with variables used in escape codes to colorize the message not defined, and there was no problem (but the colors didn't change), trying to simulate the script in Windows OS.

bobsummerwill commented:
Thanks.   Looks good to me!

Why don't you read about how to "squash" commits in git, and do that next, to collapse this PR down to a single change.   You can also "git commit --amend" on that commit, to make a detailed commit message which matches the final changes.

If you line that all up then everything is good with me.   For commits, we just need 2 of the core developers to approve.   You have my "thumbs up", so if you make the other changes then @chriseth or one of the other developers can do the merge tomorrow.

Thanks for your help and engagement!    Hopefully this can be the first of many contributions from you.

programonauta commented:
Hi, I'm a bit confused with this.
I did `git rebase -i HEAD-6` and changed the file opened with `reword` the first line with a more clear message and `squash` for others commits, following the instructions of this video: https://www.youtube.com/watch?v=qh9KtjfjzCU

After that, my git log:

```
commit 98fe53cca1086a455f2b01934871bf5fc0c35473
Merge: 69341a6 b70d402
Author: Ricardo Azevedo Brandao <ricardo@MacMiniBrandNet.home>
Date:   Sun Mar 13 12:27:34 2016 -0300

    Merge branch 'develop' of https://github.com/programonauta/webthree-umbrella into develop

commit 69341a60fbb2818484ac7e18c1d4a9a5cf3dd89b
Author: Ricardo Brandao <rbrandao.br@gmail.com>
Date:   Fri Mar 11 17:05:21 2016 -0300

    Test if sub-modules exist and give an error message if are missing

    Improvemnts and color on error message

    Test if file exists for all includes

    Testing only sub-directories

    Remove unecessary commented line

    removed escape codes not used

commit b70d4027521237c00b660194fd6ad7b03813ba83
Author: Ricardo Azevedo Brandao <ricardo@MacMiniBrandNet.home>
Date:   Sat Mar 12 21:37:52 2016 -0300

    removed escape codes not used

commit 4c0434ef350db4aa1baba5abe07462a69a39b78a
Author: Ricardo Azevedo Brandao <ricardo@MacMiniBrandNet.home>
Date:   Sat Mar 12 21:18:50 2016 -0300

    Remove unecessary commented line

commit 180f87bf323a55b751483327c16946a267ac9c1d
Author: Ricardo Azevedo Brandao <ricardo@MacMiniBrandNet.home>
Date:   Sat Mar 12 21:16:48 2016 -0300

    Testing only sub-directories

commit b4e5b0409bda74a41501b8bf9b0f85df7657f6c5
Author: Ricardo Azevedo Brandao <ricardo@MacMiniBrandNet.home>
Date:   Sat Mar 12 07:43:04 2016 -0300
```

But when I trying to git push I received an error, and I'm stuck now. :(

```
$ git push
warning: push.default is unset; its implicit value has changed in
Git 2.0 from 'matching' to 'simple'. To squelch this message
and maintain the traditional behavior, use:

  git config --global push.default matching

To squelch this message and adopt the new behavior now, use:

  git config --global push.default simple

When push.default is set to 'matching', git will push local branches
to the remote branches that already exist with the same name.

Since Git 2.0, Git defaults to the more conservative 'simple'
behavior, which only pushes the current branch to the corresponding
remote branch that 'git pull' uses to update the current branch.

See 'git help config' and search for 'push.default' for further information.
(the 'simple' mode was introduced in Git 1.7.11. Use the similar mode
'current' instead of 'simple' if you sometimes use older versions of Git)

To https://github.com/programonauta/webthree-umbrella.git
 ! [rejected]        develop -> develop (non-fast-forward)
error: failed to push some refs to 'https://github.com/programonauta/webthree-umbrella.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

I noticed that I need to set push.default. But I really don't know what to do.

programonauta commented:
I was trying to figure out and I've set push.default to simple, and pushed again:

```
$ git config --global push.default simple
$ git push
Counting objects: 2, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (2/2), done.
Writing objects: 100% (2/2), 604 bytes | 0 bytes/s, done.
Total 2 (delta 0), reused 0 (delta 0)
To https://github.com/programonauta/webthree-umbrella.git
   b70d402..98fe53c  develop -> develop
```

But as far as I could see nothing new happened

bobsummerwill commented:
Rebasing and squashing ARE a bit confusing.    Maybe you should make your own repo and do some experiments there until you have the magic worked out?   And then come back to this.

programonauta commented:
@bobsummerwill, yes it's true.
But I suppose this worked, because there is a commit and a merge branch btw my post and yours, and the commit have exactly the message I defined on rebase. And if you see the diff of this commit has only the difference between the original and the last CMakeLists.txt file.

bobsummerwill commented:
Yes - the content is correct.   The squash just turns those 8 commits into 1 commit, so that the revision history does not get forever polluted with all of these irrelevant interim steps as you iterated to the right answer.

chriseth commented:
There is still a merge commit in the history. Could you remove that by rebasing and also squashing would be a good thing.
If you say it is too complicated, we can also leave it.

programonauta commented:
@chriseth, yes I'm still trying to understand how "magic happens" using rebase :)

AFAIK, it's just run `git rebase -i HEAD~2`. Am I correct?

I really wanted to try, unless if trying to solve the problem we're under risk to have more confusion. If the solution is more difficult than that, I prefer you do it and I will create a "sandbox repository" and try to figure out how does it work.

So, please let me know what is better.

chriseth commented:
Assuming `origin` is https://github.com/ethereum/webthree-umbrella (if not, add that repository as a remote repository, the name is not too important), then do
`git fetch origin`
`git rebase -i origin/develop`.

programonauta commented:
@chriseth, trying to figure out what is origin I did:

```
$ git remote show origin
* remote origin
  Fetch URL: https://github.com/programonauta/webthree-umbrella.git
  Push  URL: https://github.com/programonauta/webthree-umbrella.git
  HEAD branch: develop
  Remote branches:
    LianaHus-patch-1 tracked
    chriseth-patch-1 tracked
    chriseth-patch-2 tracked
    crgav_test       tracked
    develop          tracked
    homestead-test   tracked
    js_location      tracked
    reflect_names    tracked
    release          tracked
  Local branch configured for 'git pull':
    develop merges with remote develop
  Local ref configured for 'git push':
    develop pushes to develop (up to date)
```

Do I need to change it? If yes, is the command `git remote set-url`?

chriseth commented:
```
git remote add ethereum https://github.com/ethereum/webthree-umbrella
git fetch ethereum
git rebase -i ethereum/develop
```

chriseth commented:
Wonderful, thanks!
